accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
 
-### New API
+There are a few options to consider for the `Assembly.Location` property of a bundled assembly:
 
-In this section, we propose adding a few APIs to facilitate common operations on bundled-apps.
+* A fixed literal (ex: `null`) indicating that no actual location is available.
+* The simple name of the assembly (with no path).
+* The path of the assembly as if it were not to be packaged into the single-file.
+* A special UNC notation such as `<bundle-path>/:/asm.dll` to denote files that come from the bundle. 
+* A configurable selection of the above, etc.
 
-The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+Proposed solution is: `:/<bundle-sub-path>/asm.dll`, where the `:` denotes the location within the bundle. The `:` is not a valid character in file-names on Windows. The `:` is already used as a path-separator on Unix systems.
 
-- Reading additional files packaged into the app (ex: data files).
-- Open an assembly for reflection/inspection
-- Load plugins built as single-file class-libs using existing Loader APIs.
+Most of the app development can be agnostic to whether the app is published as single-file or not. However, the parts of the app that deal with physical locations of files need to be aware of the single-file packaging. 
+
+#### `AppContext.BaseDirectory`
 
-Therefore, we propose adding an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netframework-4.7.2#System_Reflection_Assembly_GetManifestResourceStream_System_String_) to obtain a stream corresponding to an embedded file. 
+`AppContext.BaseDirectory` will be the directory where the AppHost (the single-file bundle itself) resides.
 
-This is only a draft of the proposed APIs. The actual shape of the APIs will be decided via API review process.
+## Handling Content 
+
+The app may want to access certain embedded content for reading, rather than loading via host/runtime. For example: bundled payload/data files. In this case, the recommended strategy is to embed the content files within appropriate managed assemblies as PE resources, and access them through [resource handling APIs](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourceinfo?view=netcore-3.1). 
+
+However, explicit access to the embedded files is useful in situations such as: 
+- Reading data-files from the app's native code
+- Open an assembly for reflection/inspection
+
+Therefore, we propose adding an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netcore-3.1) to obtain a stream corresponding to an embedded file. This is only a draft of the proposed APIs. The actual shape of the APIs will be decided via API review process.
```

Comment:
This APIs need more crisp scenarios. I do not see what this is for.

Reply:
KISS - keep it simple until we have a really good reason for needing new APIs.

Reply:
My thinking was that not requiring the content to be embedded within PE files will lead to easier adaptation, because existing applications need fewer changes. Also, the way apps are built in the non-single-file case need not change at all. 

But I agree that we can seek feedback from customers without them in previews, so I've removed the new APIs.

Reply:
Other than the content accessing APIs, I think the `IsBundle()` API (which tells whether the app is run from a single-file or not) is worth including. 

In general, apps should be agnostic to how they are published (single-file or many files). However, having the `IsBundle()` test helps customer work around situations where the single-file functionality may be missing.

Several customers have asked for this `IsBundle()` test in .net core 3.x. We've resisted from offering this API in 3.x, instead offering solutions/work-arounds for the specific customer situations -- sometimes not completely successfully. 

For example, in 3.x `AppContext.BaseDirectory` points to the extraction directory. We have given other options to locate the AppHost (ex: `Process.GetCurrentProcess().MainModule.FileName`). This works in the single-file scenario, but not if the app is run through `dotnet run` in a non-single-file publish. 

This specific situation of `AppContext.BaseDirectory` will be resolved in 5.0, but having a bundle-differentiator will likely give a chance for customers to circumvent other shortcomings until the resolved.

Reply:
> IsBundle() API

The problem with such API is that it is unclear what is the set of behaviors that it returns true for. I do not think we would be able to provide a good guidance for what to use it for, and maintain compatibility for how this API gets used as we evolve the runtime.

Reply:
My understanding of `IsBundle()` API was that it returns `true` if the host contains a bundle-header.

```cs
namespace System.Runtime.Loader
{
    public partial class Bundle
    {
        // Check whether an app is running from a single-file bundle
        public static bool IsBundle();
    }
}
```
The semantics, and the form of this API may need to evolve, along with the runtime. 
For example:
* If we have CoreRT style single-file apps, whether `IsBundle()` returns `true`? 
* If we support single-file plugins, whether we should have ``IsBundle(assembly)` ?

Reply:
> IsBundle() API was that it returns true if the host contains a bundle-header.

Given https://github.com/dotnet/designs/pull/90#discussion_r375217478 , we may end up with bundle-header for every app in .NET 5 and so it would return true all the time based on this spec.

The APIs needs to be described in terms of what people should use them for, not in terms of how they are implemented.